packages/contracts/contracts/optimistic-ethereum/bridge/L1CrossDomainMessenger.sol

```diff
@@ -0,0 +1,258 @@
+     * .inheritdoc IL1CrossDomainMessenger
+     */
+    function replayMessage(
+        address _target,
+        address _sender,
+        bytes memory _message,
+        uint256 _messageNonce,
+        uint32 _gasLimit
+    )
+        public
+    {
+        bytes memory xDomainCalldata = _getXDomainCalldata(
+            _target,
+            _sender,
+            _message,
+            _messageNonce
+        );
+
+        require(
+            sentMessages[keccak256(xDomainCalldata)] == true,
+            "Provided message has not already been sent."
+        );
+
+        _sendXDomainMessage(xDomainCalldata, _gasLimit);
+    }
+
+
+    /*
+     * Internal Functions
+     */
+
+    /**
+     * Verifies that the given message is valid.
+     * @param _xDomainCalldata Calldata to verify.
+     * @param _proof Inclusion proof for the message.
+     * @return Whether or not the provided message is valid.
+     */
+    function _verifyXDomainMessage(
+        bytes memory _xDomainCalldata,
+        L2MessageInclusionProof memory _proof
+    )
+        internal
+        returns (
+            bool
+        )
+    {
+        return (
+            _verifyStateRootProof(_proof) && _verifyStorageProof(_xDomainCalldata, _proof)
+        );
+    }
+
+    /**
+     * Verifies that the state root within an inclusion proof is valid.
+     * @param _proof Message inclusion proof.
+     * @return Whether or not the provided proof is valid.
+     */
+    function _verifyStateRootProof(
+        L2MessageInclusionProof memory _proof
+    )
+        internal
+        returns (
+            bool
+        )
+    {
+        // TODO: We *must* verify that the batch timestamp is sufficiently old.
+        // However, this requires that we first add timestamps to state batches
+        // and account for that change in various tests. Change of that size is
+        // out of scope for this ticket, so "TODO" for now.
+
+        StateCommitmentChain stateCommitmentChain = resolveStateCommitmentChain();
+        return stateCommitmentChain.verifyElement(
+            abi.encodePacked(_proof.stateRoot),
+            _proof.stateRootIndex,
+            _proof.stateRootProof
+        );
+    }
+
+    /**
+     * Verifies that the storage proof within an inclusion proof is valid.
+     * @param _xDomainCalldata Encoded message calldata.
+     * @param _proof Message inclusion proof.
+     * @return Whether or not the provided proof is valid.
+     */
+    function _verifyStorageProof(
+        bytes memory _xDomainCalldata,
+        L2MessageInclusionProof memory _proof
+    )
+        internal
+        returns (
+            bool
+        )
+    {
+        bytes32 storageKey = keccak256(
+            BytesLib.concat(
+                abi.encodePacked(keccak256(_xDomainCalldata)),
+                abi.encodePacked(uint256(0))
+            )
+        );
+
+        return EthMerkleTrie.proveAccountStorageSlotValue(
+            0x4200000000000000000000000000000000000000,
```

Comment:
This is a ovm specific precompile address? Maybe a good idea to turn it into a constant so its not a magic number